File: src/extras-karchive/libkarchiveentry.cpp
#include <KArchive>
#include <KArchiveEntry>
#include <QDateTime>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <karchiveentry.h>
#include "libkarchiveentry.h"
#include "libkarchiveentry.hxx"

KArchiveEntry* KArchiveEntry_new(KArchive* archive, const libqt_string name, int access, const QDateTime* date, const libqt_string user, const libqt_string group, const libqt_string symlink) {
    QString name_QString = QString::fromUtf8(name.data, name.len);
    QString user_QString = QString::fromUtf8(user.data, user.len);
    QString group_QString = QString::fromUtf8(group.data, group.len);
    QString symlink_QString = QString::fromUtf8(symlink.data, symlink.len);
    return new VirtualKArchiveEntry(archive, name_QString, static_cast<int>(access), *date, user_QString, group_QString, symlink_QString);
}

QDateTime* KArchiveEntry_Date(const KArchiveEntry* self) {
    return new QDateTime(self->date());
}

libqt_string KArchiveEntry_Name(const KArchiveEntry* self) {
    QString _ret = self->name();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

mode_t KArchiveEntry_Permissions(const KArchiveEntry* self) {
    return self->permissions();
}

libqt_string KArchiveEntry_User(const KArchiveEntry* self) {
    QString _ret = self->user();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KArchiveEntry_Group(const KArchiveEntry* self) {
    QString _ret = self->group();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

libqt_string KArchiveEntry_SymLinkTarget(const KArchiveEntry* self) {
    QString _ret = self->symLinkTarget();
    // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
    QByteArray _b = _ret.toUtf8();
    libqt_string _str;
    _str.len = _b.length();
    _str.data = static_cast<const char*>(malloc(_str.len + 1));
    memcpy((void*)_str.data, _b.data(), _str.len);
    ((char*)_str.data)[_str.len] = '\0';
    return _str;
}

bool KArchiveEntry_IsFile(const KArchiveEntry* self) {
    auto* vkarchiveentry = dynamic_cast<const VirtualKArchiveEntry*>(self);
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        return self->isFile();
    } else {
        return ((VirtualKArchiveEntry*)self)->isFile();
    }
}

bool KArchiveEntry_IsDirectory(const KArchiveEntry* self) {
    auto* vkarchiveentry = dynamic_cast<const VirtualKArchiveEntry*>(self);
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        return self->isDirectory();
    } else {
        return ((VirtualKArchiveEntry*)self)->isDirectory();
    }
}

void KArchiveEntry_VirtualHook(KArchiveEntry* self, int id, void* data) {
    auto* vkarchiveentry = dynamic_cast<VirtualKArchiveEntry*>(self);
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->virtual_hook(static_cast<int>(id), data);
    }
}

// Base class handler implementation
bool KArchiveEntry_QBaseIsFile(const KArchiveEntry* self) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_IsFile_IsBase(true);
        return vkarchiveentry->isFile();
    } else {
        return self->KArchiveEntry::isFile();
    }
}

// Auxiliary method to allow providing re-implementation
void KArchiveEntry_OnIsFile(const KArchiveEntry* self, intptr_t slot) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_IsFile_Callback(reinterpret_cast<VirtualKArchiveEntry::KArchiveEntry_IsFile_Callback>(slot));
    }
}

// Base class handler implementation
bool KArchiveEntry_QBaseIsDirectory(const KArchiveEntry* self) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_IsDirectory_IsBase(true);
        return vkarchiveentry->isDirectory();
    } else {
        return self->KArchiveEntry::isDirectory();
    }
}

// Auxiliary method to allow providing re-implementation
void KArchiveEntry_OnIsDirectory(const KArchiveEntry* self, intptr_t slot) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_IsDirectory_Callback(reinterpret_cast<VirtualKArchiveEntry::KArchiveEntry_IsDirectory_Callback>(slot));
    }
}

// Base class handler implementation
void KArchiveEntry_QBaseVirtualHook(KArchiveEntry* self, int id, void* data) {
    auto* vkarchiveentry = dynamic_cast<VirtualKArchiveEntry*>(self);
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_VirtualHook_IsBase(true);
        vkarchiveentry->virtual_hook(static_cast<int>(id), data);
    } else {
        ((VirtualKArchiveEntry*)self)->virtual_hook(static_cast<int>(id), data);
    }
}

// Auxiliary method to allow providing re-implementation
void KArchiveEntry_OnVirtualHook(KArchiveEntry* self, intptr_t slot) {
    auto* vkarchiveentry = dynamic_cast<VirtualKArchiveEntry*>(self);
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_VirtualHook_Callback(reinterpret_cast<VirtualKArchiveEntry::KArchiveEntry_VirtualHook_Callback>(slot));
    }
}

// Derived class handler implementation
KArchive* KArchiveEntry_Archive(const KArchiveEntry* self) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        return vkarchiveentry->archive();
    } else {
        return ((VirtualKArchiveEntry*)self)->archive();
    }
}

// Base class handler implementation
KArchive* KArchiveEntry_QBaseArchive(const KArchiveEntry* self) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_Archive_IsBase(true);
        return vkarchiveentry->archive();
    } else {
        return ((VirtualKArchiveEntry*)self)->archive();
    }
}

// Auxiliary method to allow providing re-implementation
void KArchiveEntry_OnArchive(const KArchiveEntry* self, intptr_t slot) {
    auto* vkarchiveentry = const_cast<VirtualKArchiveEntry*>(dynamic_cast<const VirtualKArchiveEntry*>(self));
    if (vkarchiveentry && vkarchiveentry->isVirtualKArchiveEntry) {
        vkarchiveentry->setKArchiveEntry_Archive_Callback(reinterpret_cast<VirtualKArchiveEntry::KArchiveEntry_Archive_Callback>(slot));
    }
}

void KArchiveEntry_Delete(KArchiveEntry* self) {
    delete self;
}

File: src/extras-karchive/libkarchiveentry.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const KArchive = @import("libqt6").KArchive;
const QDateTime = @import("libqt6").QDateTime;
const std = @import("std");

/// ### [Upstream resources](https://api.kde.org/karchiveentry.html)
pub const KArchiveEntry = extern struct {
    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html)
    ///
    /// The pointer to the underlying Qt C++ object
    ///
    ptr: QtC.KArchiveEntry,

    pub const _is_KArchiveEntry = {};

    /// ### DEPRECATED: Use `new` instead
    ///
    pub const New = new;

    /// Allocate a new KArchiveEntry object in C++ memory
    ///
    /// ## Parameter(s):
    ///
    /// ` _archive: KArchive `
    ///
    /// ` _name: []const u8 `
    ///
    /// ` access: i32 `
    ///
    /// ` _date: QDateTime `
    ///
    /// ` _user: []const u8 `
    ///
    /// ` _group: []const u8 `
    ///
    /// ` symlink: []const u8 `
    ///
    pub fn new(_archive: anytype, _name: []const u8, access: i32, _date: anytype, _user: []const u8, _group: []const u8, symlink: []const u8) KArchiveEntry {
        comptime _ = @TypeOf(_archive)._is_KArchive;
        const name_str = qtc.libqt_string{
            .len = _name.len,
            .data = _name.ptr,
        };
        comptime _ = @TypeOf(_date)._is_QDateTime;
        const user_str = qtc.libqt_string{
            .len = _user.len,
            .data = _user.ptr,
        };
        const group_str = qtc.libqt_string{
            .len = _group.len,
            .data = _group.ptr,
        };
        const symlink_str = qtc.libqt_string{
            .len = symlink.len,
            .data = symlink.ptr,
        };
        return .{ .ptr = qtc.KArchiveEntry_new(@ptrCast(_archive.ptr), name_str, @bitCast(access), @ptrCast(_date.ptr), user_str, group_str, symlink_str) };
    }

    /// ### DEPRECATED: Use `date` instead
    ///
    pub const Date = date;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#date)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn date(self: KArchiveEntry) QDateTime {
        return .{ .ptr = qtc.KArchiveEntry_Date(@ptrCast(self.ptr)) };
    }

    /// ### DEPRECATED: Use `name` instead
    ///
    pub const Name = name;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#name)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn name(self: KArchiveEntry, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KArchiveEntry_Name(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KArchiveEntry.name: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `permissions` instead
    ///
    pub const Permissions = permissions;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#permissions)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn permissions(self: KArchiveEntry) u32 {
        return @bitCast(qtc.KArchiveEntry_Permissions(@ptrCast(self.ptr)));
    }

    /// ### DEPRECATED: Use `user` instead
    ///
    pub const User = user;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#user)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn user(self: KArchiveEntry, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KArchiveEntry_User(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KArchiveEntry.user: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `group` instead
    ///
    pub const Group = group;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#group)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn group(self: KArchiveEntry, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KArchiveEntry_Group(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KArchiveEntry.group: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `symLinkTarget` instead
    ///
    pub const SymLinkTarget = symLinkTarget;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#symLinkTarget)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` allocator: std.mem.Allocator `
    ///
    pub fn symLinkTarget(self: KArchiveEntry, allocator: std.mem.Allocator) []const u8 {
        var _str = qtc.KArchiveEntry_SymLinkTarget(@ptrCast(self.ptr));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("KArchiveEntry.symLinkTarget: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// ### DEPRECATED: Use `isFile` instead
    ///
    pub const IsFile = isFile;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isFile)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn isFile(self: KArchiveEntry) bool {
        return qtc.KArchiveEntry_IsFile(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `onIsFile` instead
    ///
    pub const OnIsFile = onIsFile;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isFile)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` callback: *const fn () callconv(.c) bool `
    ///
    pub fn onIsFile(self: KArchiveEntry, callback: *const fn () callconv(.c) bool) void {
        qtc.KArchiveEntry_OnIsFile(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superIsFile` instead
    ///
    pub const SuperIsFile = superIsFile;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isFile)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn superIsFile(self: KArchiveEntry) bool {
        return qtc.KArchiveEntry_SuperIsFile(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `isDirectory` instead
    ///
    pub const IsDirectory = isDirectory;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isDirectory)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn isDirectory(self: KArchiveEntry) bool {
        return qtc.KArchiveEntry_IsDirectory(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `onIsDirectory` instead
    ///
    pub const OnIsDirectory = onIsDirectory;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isDirectory)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` callback: *const fn () callconv(.c) bool `
    ///
    pub fn onIsDirectory(self: KArchiveEntry, callback: *const fn () callconv(.c) bool) void {
        qtc.KArchiveEntry_OnIsDirectory(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superIsDirectory` instead
    ///
    pub const SuperIsDirectory = superIsDirectory;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#isDirectory)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn superIsDirectory(self: KArchiveEntry) bool {
        return qtc.KArchiveEntry_SuperIsDirectory(@ptrCast(self.ptr));
    }

    /// ### DEPRECATED: Use `archive` instead
    ///
    pub const Archive = archive;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#archive)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn archive(self: KArchiveEntry) KArchive {
        return .{ .ptr = qtc.KArchiveEntry_Archive(@ptrCast(self.ptr)) };
    }

    /// ### DEPRECATED: Use `onArchive` instead
    ///
    pub const OnArchive = onArchive;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#archive)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` callback: *const fn () callconv(.c) KArchive `
    ///
    pub fn onArchive(self: KArchiveEntry, callback: *const fn () callconv(.c) KArchive) void {
        qtc.KArchiveEntry_OnArchive(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superArchive` instead
    ///
    pub const SuperArchive = superArchive;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#archive)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn superArchive(self: KArchiveEntry) KArchive {
        return .{ .ptr = qtc.KArchiveEntry_SuperArchive(@ptrCast(self.ptr)) };
    }

    /// ### DEPRECATED: Use `virtualHook` instead
    ///
    pub const VirtualHook = virtualHook;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#virtual_hook)
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` id: i32 `
    ///
    /// ` data: ?*anyopaque `
    ///
    pub fn virtualHook(self: KArchiveEntry, id: i32, data: ?*anyopaque) void {
        qtc.KArchiveEntry_VirtualHook(@ptrCast(self.ptr), @bitCast(id), @ptrCast(data));
    }

    /// ### DEPRECATED: Use `onVirtualHook` instead
    ///
    pub const OnVirtualHook = onVirtualHook;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#virtual_hook)
    ///
    /// Allows for overriding the related default method
    ///
    /// ## Parameters:
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` callback: *const fn (self: KArchiveEntry, id: i32, data: ?*anyopaque) callconv(.c) void `
    ///
    pub fn onVirtualHook(self: KArchiveEntry, callback: *const fn (KArchiveEntry, i32, ?*anyopaque) callconv(.c) void) void {
        qtc.KArchiveEntry_OnVirtualHook(@ptrCast(self.ptr), @bitCast(@intFromPtr(callback)));
    }

    /// ### DEPRECATED: Use `superVirtualHook` instead
    ///
    pub const SuperVirtualHook = superVirtualHook;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#virtual_hook)
    ///
    /// Base class method implementation
    ///
    /// ## Parameter(s):
    ///
    /// ` self: KArchiveEntry `
    ///
    /// ` id: i32 `
    ///
    /// ` data: ?*anyopaque `
    ///
    pub fn superVirtualHook(self: KArchiveEntry, id: i32, data: ?*anyopaque) void {
        qtc.KArchiveEntry_SuperVirtualHook(@ptrCast(self.ptr), @bitCast(id), @ptrCast(data));
    }

    /// ### DEPRECATED: Use `delete` instead
    ///
    pub const Delete = delete;

    /// ### [Upstream resources](https://api.kde.org/karchiveentry.html#dtor.KArchiveEntry)
    ///
    /// Delete this object from C++ memory
    ///
    /// ## Parameter:
    ///
    /// ` self: KArchiveEntry `
    ///
    pub fn delete(self: KArchiveEntry) void {
        qtc.KArchiveEntry_Delete(@ptrCast(self.ptr));
    }
};
